packages/react-native/sdks/hermes-engine/hermes-utils.rb

```diff
@@ -177,8 +178,54 @@ def podspec_source_build_from_github_tag(react_native_path)
 end
 
 def podspec_source_build_from_github_main()
-    hermes_log("Using the latest commit from main.")
-    return {:git => HERMES_GITHUB_URL, :commit => `git ls-remote #{HERMES_GITHUB_URL} main | cut -f 1`.strip}
+    # [macOS
+    # The logic for this is a bit different on macOS.
+    # Since react-native-macos lags slightly behind react-native, we can't always use
+    # the latest Hermes commit because Hermes and JSI don't always guarantee backwards compatibility.
+    # Instead, we take the commit hash of Hermes at the time of the merge base between us and RNCore.
+
+    # We don't need ls-remote because react-native-macos is a fork of RNCore
+    fetch_result = `git fetch -q https://github.com/facebook/react-native.git`
+    if $?.exitstatus != 0
+        abort <<-EOS
+        [Hermes] Failed to fetch RNCore into the local repository.
+        EOS
+    end
+
+    merge_base = `git merge-base FETCH_HEAD HEAD`.strip
+    if merge_base.empty?
+        abort <<-EOS
+        [Hermes] Unable to find the merge base between RNCore main and the current branch.
+        EOS
+    end
+
+    timestamp = `git show -s --format=%ci #{merge_base}`.strip
+    if timestamp.empty?
+        abort <<-EOS
+        [Hermes] Unable to extract the timestamp for the merge base (#{merge_base}).
+        EOS
+    end
+
+    commit = nil
+    Dir.mktmpdir do |tmpdir|
+        hermes_git_dir = File.join(tmpdir, "hermes.git")
+        # Unfortunately we can't use git rev-list on HERMES_GITHUB_URL directly since we're not in that repo.
+        # Instead, we create a shallow clone to avoid downloading the entire history.
+        `git clone -q --bare --shallow-since="#{timestamp}" #{HERMES_GITHUB_URL} "#{hermes_git_dir}"`
+        `git --git-dir="#{hermes_git_dir}" fetch -q --deepen=1`
+
+        # If all goes well, this will be the commit hash of hermes at the time of the merge base
+        commit = `git --git-dir="#{hermes_git_dir}" rev-list -1 --before="#{timestamp}" HEAD`.strip
+        if commit.empty?
+            abort <<-EOS
+            [Hermes] Unable to find the Hermes commit hash at time #{timestamp}.
+            EOS
+        end
+    end
+
+    hermes_log("Using Hermes commit hash from the merge base with RNCore main: #{commit}")
```

Comment:
nit: prefer something like "React Native main" since RNCore is kind of an informal name we out-of-tree platforms give it.